core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java

```diff
@@ -221,9 +199,53 @@ public FormValidation doCheckUpstreamProjects(@AncestorInPath Job project, @Quer
     }
 
     @Extension public static final class RunListenerImpl extends RunListener<Run> {
+        
+        static RunListenerImpl get() {
+            return ExtensionList.lookup(RunListener.class).get(RunListenerImpl.class);
+        }
+
+        private Map<Job,Collection<ReverseBuildTrigger>> upstream2Trigger;
+
+        synchronized void invalidateCache() {
+            upstream2Trigger = null;
+        }
+
+        private Map<Job,Collection<ReverseBuildTrigger>> calculateCache() {
+            Map<Job,Collection<ReverseBuildTrigger>> result = new WeakHashMap<>();
+            SecurityContext orig = ACL.impersonate(ACL.SYSTEM);
```

Comment:
Are you sure that <code>finally</code> handler is placed correctly? It's inside for loop :ant: